Replace eggs with wheels

natefoo opened the pull request:
# Smash Eggs

More importantly, replace all of our custom egg installation code with a very slightly modified version of pip. The modification is necessary if/until [wheel is updated to include support for Linux wheels](https://bitbucket.org/pypa/wheel/pull-requests/55/soabi-support-for-python-2x-and-pypy).
## Other changes include
- Removal of the awful `lib/pkg_resources.py`
- Relocated the bit of eggs code that checked the Galaxy config to determine what dependencies are needed to a wheels module.
- [galaxyproject/docker-build](https://github.com/galaxyproject/docker-build/) has been updated with scripts and images to support building wheels for distribution on 32 and 64 bit Linux.
## On virtualenvs

`scripts/common_startup.sh` will create a virtualenv at `$galaxy_root/.venv` if one is not already active and wheels will be installed to here. Otherwise, Galaxy will use whatever virtualenv is currently active (in `$VIRTUAL_ENV`).

Whatever virtualenv Galaxy is started under will be passed to jobs if a virtualenv is not already active in the job environment (and it exists at the same path on the compute resource). You can force jobs to use a different virtualenv using the `<env>` tag in a job destination, e.g.:

``` xml
<destination id="foo">
    <env file="/path/to/venv/bin/activate"/>
</destination>
```

Galaxy needs to become installable in the traditional `python setup.py install` manner which will change some things about how virtualenvs are handled in the future (tl;dr, we don't need to "handle" them all unless you are running directly from source), but that is the next project. This was all recently done for Pulsar so Galaxy can steal from there. =)
## TODO
- [x] Replace distribute in bx-python with setuptools.
- [x] Figure out how to build OS X wheels for distribution.
- [x] Remove all eggs imports and eggs require()s
- [x] ~~Get rid of the SVGFig dependency~~
- [x] Do not attempt to install all dependencies every time (determine what's missing and attempt to install those).
- [x] Fail to start if a virtualenv does not exist and cannot be created.
- [x] Figure out how to support modified packages (Whoosh, sqlalchemy-migrate). `pip install git+https://` works, but then it's a little trickier to correctly answer "Is sqlalchemy-migrate installed, and is it the right version?"
- [x] Build [and host](https://wheels.galaxyproject.org/) wheels for everything that does not already have them (and possibly even the stuff already in PyPI). We should provide wheels for anything that does not provide them, possibly even for pure-python modules since it'll still be faster to install a wheel even though there is no C compilation involved.
- [x] Figure out how to build psycopg2 ~~and MySQL_Python~~ (anyone using MySQL can just `pip install MySQL-python` (which `common-startup.sh` will in fact attempt to do)
- [x] The pysam wheel is broken: `ImportError: /home/nate/galaxy/galaxy/.venv/local/lib/python2.7/site-packages/pysam/calignmentfile.so: undefined symbol: hts_idx_load_literal`
- [x] The bx wheel is broken (possibly due to being built w/o numpy installed?):  `ImportError: No module named bigbed_file`
- [x] Test drmaa and pbs runners
- [x] Create an `sdist` for everything on wheels.galaxyproject.org
- [x] Test and document creating (or sharing) a venv w/ necessary wheels when running on the cluster
- [x] Build remaining OS X wheels
- [x] When using `run.sh`, ignore whatever virtualenv is active and use `.venv`. If you want to use something else, either start without `run.sh` or use `run.sh --skip-venv`. In the future, we won't mess with venvs at all since Galaxy should install in the standard Pythonic way.
- [x] Merge updated SOABI code from pypa/pip#3075 into natefoo/pip linux-wheels branch
## Questions
- ~~Should we pin exact versions of dependencies? I have a strong preference to say no, and as a concession say we should pin to "less than the next major version" e.g. `SQLAlchemy<1.1`. Although this does present one small problem - we will have to ensure that we always have wheels built for every version published on PyPI (we can certainly automate this) or else pip will bypass the wheels and install from source, unless we use `pip install --only-binary`. However, if we use `--only-binary`, then we cannot trivially fall back to compiling from source on platforms that we do not have wheels for.~~ I was won over by the argument that we should pin exact versions ("exact set of versioned dependencies for an exact version of Galaxy), but that we will:
  - Fetch new dependency versions as they are released
  - Build wheels for them
  - Test Galaxy against these new wheel versions
  - If the tests pass, automatically issue a PR against Galaxy to bump the version
- ~~Should the existence of `$galaxy_root/.venv` take precedence over `$VIRTUAL_ENV` (since I keep installing all of Galaxy's wheels in my tox venv =P)?~~use `--skip-venv` if you don't want `.venv` to take precedence.

natefoo committed:
WIP: Stop using eggs, start using prebuilt wheels.

natefoo committed:
Install modified pip for fetching Linux wheels, fetch wheels from
wheels.galaxyproject.org

natefoo committed:
Remove (dist-)eggs.ini.

natefoo committed:
Fix problems in wheels.py found by flake8

mr-c commented:
Fantastic work, :+1:

dannon commented:
Regarding dependency pinning, +1 to `SQLAlchemy<1.1`.  We don't want automatic major version updates to happen without testing, but it'd definitely be nice to make minor updates less onerous.

natefoo commented:
@jxtx converted bx-python back to setuptools, so that one's done.

nsoranzo commented:
This is going to be awesome!

jmchilton commented:
`135 additions and 5,471 deletions.` :smiling_imp:

natefoo commented:
OS X wheels built with python.org builds of Python (which are multi-arch Intel builds using the 10.6 SDK) work on all later versions of OS X. I am starting to upload these to https://wheels.galaxyproject.org/

jxtx commented on scripts/common_startup.sh:
```diff
@@ -42,16 +54,46 @@ if [ ! -f $GALAXY_CONFIG_FILE ]; then
+
+
+if [ $FETCH_WHEELS -eq 1 ]; then
+    pip install git+https://github.com/natefoo/pip@linux-wheels
```
This currently checks out the repository again every time I run it. Changing to `pip install -e git+https://github.com/natefoo/pip@linux-wheels#egg=pip` allows it just to update.

natefoo committed:
Install modified pip in develop/edit mode so it won't be reinstalled
every time (courtesy @jxtx).

natefoo committed:
Add nose to requirements.txt.

nsoranzo commented:
> Remove all eggs imports and eggs require()s

@natefoo: Will it be possible to change all `pkg_resources.require()`s to `import`s?

natefoo commented:
@nsoranzo In theory it should be possible to simply remove them. They should precede the name they are importing, which should now succeed since everything should be accessible on `sys.path` from the beginning.

natefoo committed:
Remove all imports of `galaxy.eggs` and all calls to
`{eggs,pkg_resources}.require`. Probably broke some stuff where we were
trying to be smart about whether a particular egg was available, but
Galaxy does at least start.

natefoo commented:
I thought I wrote this down somewhere but now I can't find it...

The suggestion from @jxtx regarding pinning was to pin to exact versions but to automate checking for minor version updates, automatically build wheels, and then PR with the new versions so that tests are run. If tests pass, then we can merge the PR and have updated versions.

This ensures that each stable release of Galaxy will have an exact set of versioned dependencies.

natefoo commented:
@martenson Are our modifications to Whoosh forked somewhere?

natefoo commented:
bx wheel fixed with galaxyproject/docker-build@bcd773f6053d4380b417f82dfa189e96142afabe

natefoo committed:
Updating requirements.txt as I get wheels ready. TODO: automate OS X
wheels.

natefoo commented:
Also, if anyone has any ideas on how to automate building wheels on OS X, I'd appreciate them since obviously we can't use docker-build there. @jxtx claims it's possible to virtualize OS X on non-Apple hardware these days?

jxtx commented:
Definitely possible. Questionably legal. http://www.contrib.andrew.cmu.edu/~somlo/OSXKVM/

natefoo commented:
I've been working on the "figure out how to build psycopg2" option, and the plan that I like the best goes like this:

Using galaxyproject/docker-build#27 (and some stuff I still need to push to that PR) it'll be possible to build wheels on many different distros using docker. Right now I plan to build for both 32- and 64-bit Debian 6, 7, and 8, Ubuntu 12.04 and 14.04, and CentOS/RHEL 6 and 7. Also if anyone wants to add SLES/openSUSE support that would be good too. This means we get to dynamically link the vendor-provided libpq rather than our own statically-compiled and probably out-of-date-and-buggy version.

There'll also be a fallback generic `linux-x86_64` wheel that will be statically compiled against psycopg2. Maybe.

hexylena commented:
@natefoo :D

natefoo commented:
SVGFig is now available [here](https://github.com/jpivarski/svgfig), and versioned tarballs are still accessible (as of now) from Google Code via [this wiki page](https://github.com/jpivarski/svgfig/wiki/RevisionHistory). So I downloaded the upstream 1.1.6 tarball, fixed the version number in `setup.py`, created an `sdist`, uploaded that to wheels.galaxyproject.org, and then built a (generic py2) wheel from that in docker-build (and uploaded it to wheels.galaxyproject.org).

natefoo commented:
@dannon can you re-enlighten me as to what version of sqlalchemy-migrate we need to use? IIRC it was just that we needed to use the version from git as opposed to the published version in PyPI? Because the version in PyPI is only a couple weeks old now so maybe it's good.

dannon commented:
@natefoo At the time it had been EOL'd in favor of alembic, but openstack folks picked the project back up.  I'd give https://github.com/stackforge/sqlalchemy-migrate/releases/tag/0.10.0 a shot and go with that if it works.

dannon commented:
@natefoo Mentioned on IRC, but here for tracking:  the version available at the time had a pinned dependency at .9.*, but it worked fine with 1.0+, so I had removed that.  I'm assuming the new version is better.

natefoo committed:
Pin requirements

natefoo committed:
Install required dependencies from requirements.txt rather than
convoluted galaxy.wheels method. Not sure why I was doing that.
Hackathons are dangerous places.

natefoo committed:
Merge remote-tracking branch 'upstream/dev' into smash_eggs

Conflicts:
	eggs.ini
	lib/galaxy/__init__.py
	lib/galaxy/datatypes/binary.py
	lib/galaxy/datatypes/converters/bgzip.py
	lib/galaxy/datatypes/converters/gff_to_interval_index_converter.py
	lib/galaxy/datatypes/converters/interval_to_bed_converter.py
	lib/galaxy/datatypes/converters/interval_to_bedstrict_converter.py
	lib/galaxy/datatypes/converters/interval_to_coverage.py
	lib/galaxy/datatypes/converters/interval_to_fli.py
	lib/galaxy/datatypes/converters/interval_to_tabix_converter.py
	lib/galaxy/datatypes/converters/maf_to_fasta_converter.py
	lib/galaxy/datatypes/converters/maf_to_interval_converter.py
	lib/galaxy/datatypes/converters/pbed_ldreduced_converter.py
	lib/galaxy/datatypes/converters/vcf_to_interval_index_converter.py
	lib/galaxy/datatypes/converters/vcf_to_vcf_bgzip.py
	lib/galaxy/datatypes/converters/wiggle_to_simple_converter.py
	lib/galaxy/datatypes/coverage.py
	lib/galaxy/eggs/__init__.py
	lib/galaxy/eggs/dist.py
	lib/galaxy/eggs/scramble.py
	lib/galaxy/exceptions/__init__.py
	lib/galaxy/jobs/runners/pbs.py
	lib/galaxy/jobs/transfer_manager.py
	lib/galaxy/model/__init__.py
	lib/galaxy/model/migrate/check.py
	lib/galaxy/model/orm/__init__.py
	lib/galaxy/model/tool_shed_install/migrate/check.py
	lib/galaxy/web/framework/__init__.py
	lib/galaxy/web/framework/helpers/__init__.py
	lib/galaxy/web/framework/middleware/sentry.py
	lib/galaxy/web/framework/openid_manager.py
	lib/galaxy/web/framework/webapp.py
	lib/galaxy/webapps/galaxy/controllers/cloudlaunch.py
	lib/galaxy/webapps/galaxy/controllers/dataset.py
	lib/galaxy/webapps/galaxy/controllers/library_common.py
	lib/galaxy/webapps/reports/controllers/workflows.py
	lib/galaxy/webapps/tool_shed/controllers/repository.py
	lib/galaxy/webapps/tool_shed/search/repo_search.py
	lib/galaxy/webapps/tool_shed/search/tool_search.py
	lib/galaxy_ext/metadata/set_metadata.py
	lib/pulsar/client/config_util.py
	lib/tool_shed/galaxy_install/install_manager.py
	lib/tool_shed/galaxy_install/migrate/check.py
	lib/tool_shed/grids/admin_grids.py
	lib/tool_shed/grids/repository_grids.py
	lib/tool_shed/repository_types/repository_suite_definition.py
	lib/tool_shed/repository_types/tool_dependency_definition.py
	lib/tool_shed/scripts/bootstrap_tool_shed/bootstrap_util.py
	lib/tool_shed/scripts/bootstrap_tool_shed/create_user_with_api_key.py
	lib/tool_shed/scripts/check_repositories_for_functional_tests.py
	lib/tool_shed/scripts/check_s3_for_empty_tool_dependency_installation_paths.py
	lib/tool_shed/scripts/check_tool_dependency_definition_repositories.py
	lib/tool_shed/scripts/show_tool_dependency_installation_dir_contents.py
	lib/tool_shed/util/basic_util.py
	lib/tool_shed/util/encoding_util.py
	lib/tool_shed/util/hg_util.py
	scripts/check_eggs.py
	scripts/dist-scramble.py
	scripts/fetch_eggs.py
	scripts/make_egg_packager.py
	scripts/scramble.py
	scripts/scramble/lib/scramble_lib.py
	scripts/scramble/scripts/DRMAA_python-macosx.py
	scripts/scramble/scripts/DRMAA_python-solaris.py
	scripts/scramble/scripts/DRMAA_python.py
	scripts/scramble/scripts/MySQL_python-solaris.py
	scripts/scramble/scripts/MySQL_python.py
	scripts/scramble/scripts/PyRods.py
	scripts/scramble/scripts/bx_python-solaris.py
	scripts/scramble/scripts/generic.py
	scripts/scramble/scripts/pbs_python.py
	scripts/scramble/scripts/psycopg2-linux.py
	scripts/scramble/scripts/psycopg2-macosx.py
	scripts/scramble/scripts/psycopg2-solaris.py
	scripts/scramble/scripts/psycopg2.py
	scripts/scramble/scripts/pysam-solaris.py
	scripts/scramble/scripts/pysqlite.py
	scripts/test_dist_egg.py

natefoo committed:
Fix linting errors

natefoo committed:
Remove egg cruft from scripts/

natefoo committed:
tox: Install nose and NoseHTML for unit tests.

natefoo committed:
Fix additional lint errors created by modifying stuff in scripts/

natefoo committed:
Fix ctabix import errors (would these converters ever have worked?)

natefoo committed:
Conditional import for irods

natefoo committed:
Conditional imports for optional dependencies.

natefoo committed:
Egg cleaning in test/

natefoo committed:
Rename galaxy.wheels to the more generalized galaxy.dependencies (since
technically we don't care whether they are wheels), handle optional deps
more correctly. Add error handling to common_startup.sh.

natefoo committed:
Fixes for job_conf.xml conditional dependencies

natefoo committed:
Remove pkg_resources.requires from lib/

natefoo committed:
Add missing WebError requirement

natefoo committed:
Add Pygments as a dependency of WebError and update SQLAlchemy version

natefoo committed:
Update Mercurial to 3.5

natefoo committed:
pysqlite is not a dependency (we expect Python's sqlite module to exist)

natefoo committed:
Handle the case where there are no conditional (heh) requirements

natefoo committed:
psycopg2 is not a hard requirement

natefoo committed:
Slurm runner is a subclass of DRMAA so its use should trigger the
conditional dependency code for the drmaa wheel.

natefoo committed:
Pass Galaxy's virtualenv to jobs (local and remote) if a venv is not
set. Can be overridden with <env file="/path/to/activate"/> on the
destination.

natefoo committed:
Fix flake8 error

natefoo committed:
Fix broken unit test caused by addition of the job wrapper's
galaxy_virtual_env attribute

natefoo committed:
Allow the use of common_startup.sh's --skip-eggs from run_tests.sh, and
have tox pass this flag (so that dependencies are installed in the tox
envs)

natefoo committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into smash_eggs

natefoo commented:
WIP tags removed!

I don't think it is ready for a merge just yet, but I'd love to see some testing on whatever crazy environments you can come up with. Nice things to test:
- Non-Debianish Linuxes
- Macs of various flavors
- Preexisting Galaxy environments using eggs switching to wheels
- Fresh clones never before run
- Inside a virtualenv, or not
- Inside a virtualenv and using `--skip-venv`
- Galaxy servers using PostgreSQL on a variety of platforms
- Upgrading a CloudMan instance
- Installing your own wheels to a venv using `pip install --index-url=https://wheels.galaxyproject.org/ -r requirements.txt` and starting `python ./scripts/paster.py` by hand. You will need to install pip w/ SOABI support first, with: `pip install --pre --no-index --find-links https://wheels.galaxyproject.org/simple/pip --upgrade pip`
- Using a non-standard Python (compiled from source)
- Run Galaxy on your PDP-11

martenson commented:
fresh start (clone and `/.run.sh`) on OSX10.9.5 seems ok, promising!

nsoranzo commented on scripts/common_startup.sh:
```diff
@@ -43,16 +55,48 @@ if [ ! -f $GALAXY_CONFIG_FILE ]; then
+    if [ ! -d .venv ]
+    then
+        vvers=13.1.0
+        vurl="https://pypi.python.org/packages/source/v/virtualenv/virtualenv-${vvers}.tar.gz"
```
This 2 lines can be moved under the `else` below.

natefoo committed:
Move vvers and vurl vars to a more relevant scope.

nsoranzo commented on lib/pulsar/client/config_util.py:
```diff
@@ -58,9 +58,6 @@ def __find_type(path, explicit_type, default_type):
 
```
You had a wrong merge with `dev` on this file, you are missing this: https://github.com/natefoo/galaxy/commit/033b1e88da99b86ffbd33321374aa59d1d72484c#diff-8ce0f1f93264f339e0bf364c7c19d3bdL4

natefoo replied:
Seriously amazing, how do you catch this stuff? =)

nsoranzo replied:
Reading line-by-line :bird: :eyes: 

This file had no `from galaxy import eggs` removed, so it looked suspicious.

natefoo committed:
Fix a bad merge in pulsar client lib caught (as always) by @nsoranzo

nsoranzo commented on lib/pulsar/client/config_util.py:
```diff
@@ -58,9 +52,6 @@ def __find_type(path, explicit_type, default_type):
 def __read_yaml(path):
-    if yaml is None:
-        raise ImportError("Attempting to read YAML configuration file - but PyYAML dependency unavailable.")
-
```
Should these lines be restored, now that `import yaml` is in a `try-except`?

natefoo committed:
Missed the rest of the bad merge in the previous commit

nsoranzo commented on lib/galaxy/util/plugin_config.py:
```diff
@@ -78,8 +69,5 @@ def plugin_source_from_path(path):
 def __read_yaml(path):
-    if yaml is None:
-        raise ImportError("Attempting to read YAML configuration file - but PyYAML dependency unavailable.")
-
```
I guess the changes to this file should be reverted (apart from the first try-except removal) to be coherent with `lib/pulsar/client/config_util.py` .

nsoranzo commented on lib/galaxy/objectstore/s3_multipart_upload.py:
```diff
@@ -14,17 +14,8 @@
-except ImportError:
-    boto = None
+import boto
+from boto.s3.connection import S3Connection
```
`boto` is not in `lib/galaxy/dependencies/requirements.txt` , so maybe it would be better to leave the try-except here with a `assert boto is not None, BOTO_IMPORT_MESSAGE` below?

jmchilton replied:
@natefoo - Yeah, even if we made a boto a required dependency there is no harm in letting something like the object store module be usable without it. By the same token yaml is no longer an optional dependency in pulsar, but I still like the `ImportError` and throw a more specific message at startup time thing in case we ever want to reuse portions of the code in different places.

natefoo replied:
Switching it back (in fact, `galaxy.objectstore.s3` already handles this correctly, but it probably would not have worked if you actually loaded the S3 object store, since `.s3_multipart_upload` was imported before `boto`). It doesn't look like we should need a `if boto is None` handler in here since that will already be done by the S3 object store.

nsoranzo commented on lib/galaxy/dependencies/conditional_requirements.txt:
```diff
@@ -0,0 +1,10 @@
+raven
+pbs_python
+drmaa
+# PyRods not in PyPI
```
Should `boto` be in this list?

natefoo replied:
Probably, but I guess we'll have to parse the entire `object_store_conf.xml` to find any configured S3 stores? The error message in `s3.py` is thankfully pretty helpful

nsoranzo replied:
Also `poster`, `python_daemon`, `requests` and `twill`?

natefoo replied:
- `boto`: Added the conditional for this only to discover that it's already a dependency of non-optional `bioblend`, so I added it to `requirements.txt` instead.
- `poster`: Is not in Pulsar's `requirements.txt` or `setup.py` yet seems to be required if not using `pycurl` or `requests` transport. @jmchilton ?
- `requests`: Pulsar will gracefully fall back to `poster`+`urllib/urllib2` (use of `requests` is not explicitly configurable).
- `python_daemon`: I think it's pretty unlikely anyone is actually using the transfer manager, which is the only thing that uses python_daemon.
- `twill`: Automatically installed into the tox venvs by tox, not needed for Galaxy otherwise.

nsoranzo replied:
`requests>=2.4.3` is also a requirement for `bioblend`, it so should be probably added to `requirements.txt` .

natefoo replied:
Added, thanks as always!

natefoo committed:
More graceful handling of failure to import yaml (a la Pulsar)

natefoo committed:
Conditional boto import in s3_multipart_upload.py

natefoo committed:
Dockerized tests need to create and activate .venv if it does not exist.
Do this by calling common_startup.sh prior to manage_db.sh in
run_test_wrapper.sh. Will require a rebuild of the testing-base image.

Also, have scripts in Galaxy root activate .venv if it exists.

martenson commented:
@natefoo you can request new tests by commenting `@galaxybot test this`

natefoo committed:
Add boto as an explicit requirement (as a dependency of bioblend)

martenson commented:
@natefoo Once this gets rebased or dev gets merged-in we can test this again.

natefoo committed:
Merge branch 'dev' into smash_eggs

natefoo committed:
Bump docker test image to 15.10.1

natefoo commented:
@galaxybot test this

natefoo committed:
Add requests as a dependency of bioblend

natefoo committed:
Make galaxy.dependencies env-config-var-aware.

natefoo commented:
Another update to the docker image is needed, see #945.

natefoo committed:
Rename conditional_requirements.txt ('_' -> '-'), add
dev-requirements.txt, fix a flake8 violation that doesn't violate PEP-8.
=P

martenson commented:
@galaxybot test this

martenson commented:
output of -framework tests

```
functional_tests.py ERROR 2015-10-20 15:47:18,153 Failure running tests
Traceback (most recent call last):
  File "./scripts/functional_tests.py", line 492, in main
    success = _run_functional_test( )
  File "./scripts/functional_tests.py", line 441, in _run_functional_test
    import functional.test_toolbox
  File "/galaxy/test/functional/test_toolbox.py", line 3, in <module>
    from base.twilltestcase import TwillTestCase
  File "/galaxy/test/base/twilltestcase.py", line 21, in <module>
    import twill
ImportError: No module named twill
```

natefoo commented:
@martenson That will be fixed when #945 is merged and a new testing image is built.

natefoo committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into smash_eggs

natefoo commented:
@galaxybot test this

natefoo committed:
Add lxml as a dev requirement for twill

natefoo committed:
Downgrade twill to 0.9.1 so we don't have to install lxml, which won't
compile because libxml-dev isn't installed in the testing image, so we'd
have to rebuild it yet again...

natefoo commented:
So close. :cry:

martenson commented:
this is reliable failure: new occurence at https://jenkins.galaxyproject.org/job/docker-framework/628/testReport/

natefoo committed:
Update the BAM datatype for pysam API changes. index_file also no longer
exists (the index must be filename + b".bai" or filename[:-4] + b".bai")
=/.  Removing it here is okay since the impact is that we can't fetch()
random regions or use .pileup(), neither of which are done here. But it
might be a problem for galaxy.visualization.data_providers.genome.

natefoo committed:
Talked to Dan, who corrected me on a few things. Also, forgot that the
whole reason we use a modified pysam is for the index_filename param...

natefoo committed:
Updates to vis data providers for pysam API changes

natefoo commented:
![IT'S HAPPENING](https://i.imgur.com/7drHiqr.gif)

martenson commented:
damn @natefoo ; I think @jxtx does not approve of that emoticon :D

gratz on the tests though, wheels are coming!

hexylena commented:
@natefoo I can't find a reaction gif for how excited I am for this. Well done.

natefoo commented:
So on a scale of -1 to +1, how do people feel about this? =D

jxtx commented:
![image](https://usatlife.files.wordpress.com/2015/02/tumblr_nch4z4zxnv1sjav2eo2_250.gif?w=1000)

martenson commented:
:+1:

jmchilton commented:
This is going to break a lot of stuff, I would suggest we get drunk tonight
and merge this at Zeno's. #mergethebigonesdrunk
On Oct 22, 2015 12:53 PM, "Eric Rasche" notifications@github.com wrote:

> @natefoo https://github.com/natefoo I can't find a reaction gif for how
> excited I am for this. Well done.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/galaxyproject/galaxy/pull/428#issuecomment-150283561.

hexylena commented:
Works for me on:
- pre-existing with eggs
- intside a venv and not

Doesn't work for me on:
- virtualenv using --skip-venv
  
  ```
  (asdf)hxr@leda:~/work/galaxy$ sh run.sh --skip-venv
  Activating virtualenv at /home/hxr/work/galaxy/.venv
  Obtaining pip from git+https://github.com/natefoo/pip@linux-wheels#egg=pip
  Updating ./.venv/src/pip clone (to linux-wheels)
  ```
  
  that flag seems to only be available on run_tests.sh
- Installing own wheels to a venv using pip install...
  
  ```
  (asdf)hxr@leda:~/work/galaxy$ pip install --index-url=https://wheels.galxayproject.org/ -r requirements.txt
  Downloading/unpacking Cheetah==2.4.4 (from -r requirements.txt (line 2))
  Cannot fetch index base URL https://wheels.galxayproject.org/
  Could not find any downloads that satisfy the requirement Cheetah==2.4.4 (from -r requirements.txt (line 2))
  Cleaning up...
  No distributions at all found for Cheetah==2.4.4 (from -r requirements.txt (line 2))
  Storing debug log for failure in /home/hxr/.pip/pip.log
  ```

![1335497891219](https://cloud.githubusercontent.com/assets/458683/10672232/522c985c-78b4-11e5-8fe4-f795c0b24f0a.jpg)

Untested:
- Non-Debianish Linuxes
- Macs of various flavors
- Galaxy servers using PostgreSQL on a variety of platforms
- Upgrading a CloudMan instance
- Using a non-standard Python (compiled from source)
- Run Galaxy on your PDP-11

hexylena commented:
That said, neither of those cases are things which should prevent this merge and can't be solved in the next two months... so :+1: 
![00](https://cloud.githubusercontent.com/assets/458683/10672266/8650d954-78b4-11e5-861a-a5c24d2a2277.jpg)

natefoo commented:
@erasche Why fix later what I can fix now before I forget how everything works?

@jmchilton +:100:

dannon commented:
Super exciting, but, for me on OSX, upgrading from existing install:

```
Found existing installation: mercurial 3.5.2
Detected a distutils installed project ('mercurial') which we cannot uninstall. The metadata provided by distutils does not contain a list of files which have been installed, so pip does not know which files to uninstall.
You must give at least one requirement to install (see "pip help install")
Activating virtualenv at /Users/yoplait/work/galaxy/.venv
Traceback (most recent call last):
  File "./scripts/paster.py", line 26, in <module>
    from galaxy.util.pastescript import serve
  File "/Users/yoplait/work/galaxy/lib/galaxy/util/__init__.py", line 36, in <module>
    import docutils.core
ImportError: No module named docutils.core
```

Wiped out pre-existing .venv, same thing.

natefoo commented:
@dannon What's your `$PYTHONPATH` look like? virtualenv-burrito screws everything up.

dannon commented:
```
[yoplait@malleus:~/work/galaxy] echo $PYTHONPATH                               smash_eggs ✭
/usr/local/lib/python2.7/site-packages
```

(brew installed python on El Capitan!)

hexylena commented:
@natefoo 

> Why fix later what I can fix now before I forget how everything works?

fair enough. :)

natefoo commented:
@erasche Well why'd you type `wheels.galxay` instead of `wheels.galaxy`... oh... nevermind. (fixed my comment above where I typo'd it). Also, `--skip-venv` support coming shortly.

@dannon What if you start with `PYTHONPATH` unset?

If it works, that's gonna be a fun one to deal with. Perhaps we should `unset PYTHONPATH` in the event that `--skip-venv` is not used.

hexylena commented:
@natefoo tested again with a correct URL, works. :) 

Edit:

```
Downloading/unpacking python-lzo==1.8 (from -r requirements.txt (line 10))
  Could not find any downloads that satisfy the requirement python-lzo==1.8 (from -r requirements.txt (line 10))
Cleaning up...
No distributions at all found for python-lzo==1.8 (from -r requirements.txt (line 10))
Storing debug log for failure in /home/hxr/.pip/pip.log
```

natefoo committed:
Use proper PEP-440 local version identifiers for modified dependencies.

natefoo committed:
Install our modified pip as a wheel from wheels.galaxyproject.org rather
than an editable directly from github.

Also, use `set -e` in scripts/common_startup.sh to fail if anything
fails, rather than chaining a bunch of things with `&&` but not properly
handling failure cases.

natefoo committed:
Pass relevant args from run.sh to common_startup.sh.

dannon commented:
@natefoo With an unset pythonpath, there is no mercurial issue, but then I hit:

```
  Could not find a version that satisfies the requirement pysam==0.8.3gx1 (from -r requirements.txt (line 13)) (from versions: 0.8.3+gx1)
No matching distribution found for pysam==0.8.3gx1 (from -r requirements.txt (line 13))
You must give at least one requirement to install (see "pip help install")
Activating virtualenv at /Users/yoplait/work/galaxy/.venv
Traceback (most recent call last):
  File "./scripts/paster.py", line 26, in <module>
    from galaxy.util.pastescript import serve
  File "/Users/yoplait/work/galaxy/lib/galaxy/util/__init__.py", line 36, in <module>
    import docutils.core
ImportError: No module named docutils.core
```

nturaga commented:
`Activating virtualenv at /Users/nturaga/Documents/PR_testing/galaxy/.venv
Ignoring indexes: https://pypi.python.org/simple
Requirement already up-to-date: pip in ./.venv/lib/python2.7/site-packages
Requirement already satisfied (use --upgrade to upgrade): Cheetah==2.4.4 in ./.venv/lib/python2.7/site-packages (from -r requirements.txt (line 2))
...
serving on http://127.0.0.1:8080`

Works for me...

natefoo commented:
For the record, @erasche's problems were fixed with the addition of installing the Galaxy-modified `pip`.

natefoo committed:
Unset $PYTHONPATH in run.sh and common_startup.sh unless --skip-env is
provided.

dannon commented:
+1
